### PR TITLE
Adds custom filtering option to alignments tracks

### PR DESCRIPTION
**Pull request by carolinebridge-oicr**
- moves filter interface definition to shared file
- adds optional custom filters to snp and pileup displays

supports https://github.com/GMOD/Apollo3/issues/233

### CODE REVIEW

File: plugins/alignments/src/LinearPileupDisplay/SharedLinearPileupDisplayMixin.ts [hunk 1]
```diff
@@ -97,6 +91,10 @@ export function SharedLinearPileupDisplayMixin(
          * #property
          */
         filterBy: types.optional(FilterModel, {}),
+        /**
+         * #property
+         */
+        customFilters: types.optional(types.array(types.string), []),
```
Comment (cmdcolin) [inline plugins/alignments/src/LinearPileupDisplay/SharedLinearPileupDisplayMixin.ts:97]: maybe jexlFilters?

**Comment by cmdcolin** [comment]
this is a random note but having the ability to arbitrarily apply some filters to all tracks may be useful, so i could forsee these arbitrary jexl filters "moving up the stack" to base display or similar https://github.com/GMOD/jbrowse-components/issues/1988

**Comment by carolinebridge-oicr** [comment]
> this is a random note but having the ability to arbitrarily apply some filters to all tracks may be useful, so i could forsee these arbitrary jexl filters "moving up the stack" to base display or similar #1988

let me know if you'd like me to do this refactor as a part of this PR -- shouldn't be complicated, but not sure if you were totally convinced of it in this comment

File: plugins/alignments/src/LinearPileupDisplay/SharedLinearPileupDisplayMixin.ts [hunk 2]
```diff
@@ -230,9 +228,18 @@ export function SharedLinearPileupDisplayMixin(
...
        */
-      setFilterBy(filter: Filter) {
+      setFilterBy(filter: IFilter) {
         self.filterBy = cast(filter)
       },
+
+      /**
+       * #action
+       */
+      setJexlFilters(filters: string[]) {
```
Comment (cmdcolin) [inline plugins/alignments/src/LinearPileupDisplay/SharedLinearPileupDisplayMixin.ts:238]: technically this is only able to push new things to the jexlFilters array, and is not able to clear it. so it's more like an "addToJexlFilters". Can probably just say self.jexlFilters=filters though to fix, or self.jexlFilters=cast(filters) if typescript complains

**Comment by cmdcolin** [comment]
we can probably keep it in the alignments track realm and move up stack later if it makes sense

**Comment by cmdcolin** [comment]
if it was possible to create a demo for this and/or integration test, it may be useful. it would help jbrowse from accidentally breaking the behavior that apollo relies on for example.

**Comment by carolinebridge-oicr** [comment]
@cmdcolin added a small test for setting the jexl filters -- not sure if that's what you meant, but accessing that property is what was described to me as how apollo is going to use it

Apollo should similarly have an e2e test on this property to cover that

**Comment by cmdcolin** [comment]
just a note: in https://github.com/GMOD/Apollo3/issues/233 it is mentioned that there is a desire to filter features that have an "edge" at a particular location. these edges may not be the start or end of a feature, but instead, indicated by the splicing from the CIGAR string. possibly an advanced jexl function could be used to satisfy this purpose

**Review by garrettjstevens — APPROVED** [review]
Looks good, I think with this that Apollo can add filters (maybe with custom jexl functions) as it needs.

**Pull request merged**